Userland/Applications/FileManager/FileUtils.cpp

```diff
@@ -67,4 +68,91 @@ void delete_paths(const Vector<String>& paths, bool should_confirm, GUI::Window*
+{
+    auto trash_bin_path = String::formatted("{}/.trash", getenv("HOME"));
+    auto trash_bin_files = String::formatted("{}/files", trash_bin_path);
+    auto trash_bin_info = String::formatted("{}/info", trash_bin_path);
+
+    if (Core::File::is_directory(trash_bin_path)) {
+        dbgln("Trash directory exists.");
+    } else {
+        if (mkdir(trash_bin_path.characters(), 0755) < 0) {
+            auto saved_errno = errno;
+            warnln("Failed to create directory {}: {}", trash_bin_path, strerror(saved_errno));
+        }
+    }
+
+    if (!Core::File::is_directory(trash_bin_files)) {
+        if (mkdir(trash_bin_files.characters(), 0755) < 0) {
+            auto saved_errno = errno;
+            warnln("Failed to create directory {}: {}", trash_bin_files, strerror(saved_errno));
+        }
+    }
+
+    if (!Core::File::is_directory(trash_bin_info)) {
+        if (mkdir(trash_bin_info.characters(), 0755) < 0) {
+            auto saved_errno = errno;
+            warnln("Failed to create directory {}: {}", trash_bin_info, strerror(saved_errno));
+        }
+    }
+}
+
+[[maybe_unused]] void move_to_trash(String const& path)
+{
+    [[maybe_unused]] auto file_info = "[Trash Info]\nPath={}\nDeletionDate={}\n";
+
+    auto file_name = LexicalPath(path).basename();
+    auto file_info_name = String::formatted("{}.trashinfo", file_name);
+    auto deletion_date = Core::DateTime::now().to_string("%Y-%m-%dT%H:%M:%S");
+
+    auto trash_bin_path = String::formatted("{}/.trash", getenv("HOME"));
+    auto trash_bin_files = String::formatted("{}/files", trash_bin_path);
+    auto trash_bin_info = String::formatted("{}/info", trash_bin_path);
+
+    create_trash_bin();
+
+    auto trash_into_path = String::formatted("{}/{}", trash_bin_info, file_info_name);
+
+    auto trashed_file_info = String::formatted(file_info, path, deletion_date);
+
+    auto file_or_error = Core::File::open(trash_into_path, Core::IODevice::WriteOnly);
+    if (file_or_error.is_error())
+        return;
+    auto file = file_or_error.release_value();
+    file->write(trashed_file_info);
+    file->close();
+
+    auto trash_file_path = String::formatted("{}/{}", trash_bin_files, file_name);
+
+    auto rc = rename(path.characters(), trash_file_path.characters());
+
+    if (rc < 0) {
+        if (errno == EXDEV) {
+            auto result = Core::File::copy_file_or_directory(
+                trash_file_path,
+                path,
+                Core::File::RecursionMode::Allowed,
+                Core::File::LinkMode::Allowed,
+                Core::File::AddDuplicateFileMarker::No);
+
+            if (result.is_error()) {
+                dbgln("Could not move '{}' to trash: {}", path, result.error().error_code);
+                return;
+            } else {
+                rc = unlink(path.characters());
+                if (rc < 0)
+                    dbgln("Could not unlink '{}': {}", path, strerror(errno));
+            }
+        }
+    }
```

Comment:
These nested conditionals contain `else` after `return`.

Also, what happens if `rc` *is* set, but `errono` is *not* `EXDEV` ? Is that possible?

Also, the outer block has no `else`, so you could return early before performing a bunch of unnecessary nested error checking.

```suggestion
    if (rc)
        return;

    if (errno == EXDEV) {
        auto result = Core::File::copy_file_or_directory(
            trash_file_path,
            path,
            Core::File::RecursionMode::Allowed,
            Core::File::LinkMode::Allowed,
            Core::File::AddDuplicateFileMarker::No);

        if (result.is_error()) {
            dbgln("Could not move '{}' to trash: {}", path, result.error().error_code);
            return;
        }

        rc = unlink(path.characters());
        if (rc < 0)
            dbgln("Could not unlink '{}': {}", path, strerror(errno));
    } else {
        // rename failed, and errno is not EXDEV
        // ...
    }
```